### PR TITLE
refactor(stream): remove executor v1

### PR DESCRIPTION
**Pull request by wangrunji0408**
## What's changed and what's your intention?

Now that all the streaming executors have been migrated to v2, we can start cleaning up the legacy v1 code.

This PR:
- removes executor v1 and v1-v2 compatibility layer.
- moves remaining code (e.g. `Barrier`, `Message`) from `executor` to `executor_v2` module.
- leaves only executor builders in the `executor` module.

Next, I plan to:
- move all executor builders in v2 to the `executor` module and rename it to `builder`.
- rename `executor_v2` to `executor`.
- make this crate's rustdoc looks better.

Any suggestions for this plan? 👀

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

## Refer to a related PR or issue link (optional)

#1392

### CODE REVIEW

File: src/stream/src/executor_v2/mod.rs [hunk 1]
```diff
@@ -129,3 +156,309 @@ pub trait Executor: Send + 'static {
...
+    pub prev: u64,
+}
+
+impl Epoch {
+    pub fn new(curr: u64, prev: u64) -> Self {
+        assert!(curr > prev);
+        Self { curr, prev }
+    }
+
+    pub fn inc(&self) -> Self {
```
Comment (skyzh) [inline src/stream/src/executor_v2/mod.rs:186]: Should rename it to something like `inc_in_test()` or mark it as `cfg(test)`. Epochs don't have such properties generally.

**Review by skyzh — APPROVED** [review]
LGTM! Please resolve conflicts. Also I guess @xxchan is modifying barrier-related logics. May discuss who to merge first :) https://github.com/singularity-data/risingwave/pull/2041/files

**Comment by skyzh** [comment]
CI failed because of https://github.com/singularity-data/risingwave/issues/2219, may re-run to pass CI.

**Review by xxchan — COMMENTED** [review]
> Also I guess @xxchan is modifying barrier-related logics. May discuss who to merge first :)

This PR can go first because I'm stuck :)

**Comment by wangrunji0408** [comment]
This PR caused an e2e test to fail: https://github.com/singularity-data/risingwave/runs/6225986559?check_suite_focus=true#step:16:151.
I'm now debugging. 🤪

**Comment by skyzh** [comment]
```
thread 'tokio-runtime-worker' panicked at 'index out of bounds: the len is 5 but the index is 6', src/stream/src/executor_v2/managed_state/join/mod.rs:144:24
stack backtrace:
   0: rust_begin_unwind
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/core/src/panicking.rs:142:14
   2: core::panicking::panic_bounds_check
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/core/src/panicking.rs:84:5
   3: <usize as core::slice::index::SliceIndex<[T]>>::index
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/core/src/slice/index.rs:242:10
   4: core::slice::index::<impl core::ops::index::Index<I> for [T]>::index
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/core/src/slice/index.rs:18:9
   5: <alloc::vec::Vec<T,A> as core::ops::index::Index<I>>::index
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/alloc/src/vec/mod.rs:2578:9
   6: risingwave_stream::executor_v2::managed_state::join::JoinHashMap<K,S>::new::{{closure}}
             at ./src/stream/src/executor_v2/managed_state/join/mod.rs:144:24
   7: core::iter::adapters::map::map_fold::{{closure}}
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/core/src/iter/adapters/map.rs:84:28
   8: core::iter::traits::iterator::Iterator::fold
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/core/src/iter/traits/iterator.rs:2362:21
   9: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::fold
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/core/src/iter/adapters/map.rs:124:9
  10: core::iter::traits::iterator::Iterator::for_each
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/core/src/iter/traits/iterator.rs:779:9
  11: <alloc::vec::Vec<T,A> as alloc::vec::spec_extend::SpecExtend<T,I>>::spec_extend
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/alloc/src/vec/spec_extend.rs:40:17
  12: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter_nested::SpecFromIterNested<T,I>>::from_iter
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/alloc/src/vec/spec_from_iter_nested.rs:62:9
  13: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/alloc/src/vec/spec_from_iter.rs:33:9
  14: <alloc::vec::Vec<T> as core::iter::traits::collect::FromIterator<T>>::from_iter
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/alloc/src/vec/mod.rs:2599:9
  15: core::iter::traits::iterator::Iterator::collect
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/core/src/iter/traits/iterator.rs:1784:9
  16: itertools::Itertools::collect_vec
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/itertools-0.10.3/src/lib.rs:1985:9
  17: risingwave_stream::executor_v2::managed_state::join::JoinHashMap<K,S>::new
             at ./src/stream/src/executor_v2/managed_state/join/mod.rs:142:29
  18: risingwave_stream::executor_v2::hash_join::HashJoinExecutor<K,S,_>::new
             at ./src/stream/src/executor_v2/hash_join.rs:383:21
  19: <risingwave_stream::executor_v2::hash_join::HashJoinExecutorDispatcher<S,_> as risingwave_common::hash::dispatcher::HashKeyDispatcher>::dispatch
             at ./src/stream/src/executor_v2/hash_join.rs:152:21
  20: risingwave_common::hash::dispatcher::HashKeyDispatcher::dispatch_by_kind
             at ./src/common/src/hash/dispatcher.rs:58:35
  21: <risingwave_stream::executor_v2::hash_join::HashJoinExecutorBuilder as risingwave_stream::executor::ExecutorBuilder>::new_boxed_executor::create_hash_join_executor
             at ./src/stream/src/executor_v2/hash_join.rs:213:63
  22: <risingwave_stream::executor_v2::hash_join::HashJoinExecutorBuilder as risingwave_stream::executor::ExecutorBuilder>::new_boxed_executor
             at ./src/stream/src/executor_v2/hash_join.rs:258:9
  23: risingwave_stream::executor::create_executor
             at ./src/stream/src/executor/mod.rs:83:5
  24: risingwave_stream::task::stream_manager::LocalStreamManagerCore::create_nodes_inner
             at ./src/stream/src/task/stream_manager.rs:509:24
  25: risingwave_stream::task::stream_manager::LocalStreamManagerCore::create_nodes
             at ./src/stream/src/task/stream_manager.rs:528:13
  26: risingwave_stream::task::stream_manager::LocalStreamManagerCore::build_actors
             at ./src/stream/src/task/stream_manager.rs:626:17
  27: risingwave_stream::task::stream_manager::LocalStreamManager::build_actors
             at ./src/stream/src/task/stream_manager.rs:290:9
  28: <risingwave_compute::rpc::service::stream_service::StreamServiceImpl as risingwave_pb::stream_service::stream_service_server::StreamService>::build_actors::{{closure}}
             at ./src/compute/src/rpc/service/stream_service.rs:65:19
  29: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/core/src/future/mod.rs:91:19
  30: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/core/src/future/future.rs:124:9
  31: <<risingwave_pb::stream_service::stream_service_server::StreamServiceServer<T> as tower_service::Service<http::request::Request<B>>>::call::BuildActorsSvc<T> as tonic::server::service::UnaryService<risingwave_pb::stream_service::BuildActorsRequest>>::call::{{closure}}
             at ./src/prost/src/stream_service.rs:548:63
  32: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/core/src/future/mod.rs:91:19
  33: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/core/src/future/future.rs:124:9
  34: tonic::server::grpc::Grpc<T>::unary::{{closure}}
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tonic-0.7.1/src/server/grpc.rs:199:27
  35: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/core/src/future/mod.rs:91:19
  36: <risingwave_pb::stream_service::stream_service_server::StreamServiceServer<T> as tower_service::Service<http::request::Request<B>>>::call::{{closure}}
             at ./src/prost/src/stream_service.rs:565:58
  37: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/core/src/future/mod.rs:91:19
  38: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/core/src/future/future.rs:124:9
  39: <F as futures_core::future::TryFuture>::try_poll
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-core-0.3.21/src/future.rs:82:9
  40: <futures_util::future::try_future::into_future::IntoFuture<Fut> as core::future::future::Future>::poll
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.21/src/future/try_future/into_future.rs:34:9
  41: <futures_util::future::future::map::Map<Fut,F> as core::future::future::Future>::poll
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.21/src/future/future/map.rs:55:37
  42: <futures_util::future::future::Map<Fut,F> as core::future::future::Future>::poll
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.21/src/lib.rs:91:13
  43: <futures_util::future::try_future::MapOk<Fut,F> as core::future::future::Future>::poll
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.21/src/lib.rs:91:13
  44: <tower::util::map_response::MapResponseFuture<F,N> as core::future::future::Future>::poll
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tower-0.4.12/src/macros.rs:38:17
  45: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/core/src/future/future.rs:124:9
  46: <tower::util::oneshot::Oneshot<S,Req> as core::future::future::Future>::poll
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tower-0.4.12/src/util/oneshot.rs:97:38
  47: <axum::routing::route::RouteFuture<B,E> as core::future::future::Future>::poll
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/axum-0.5.4/src/routing/route.rs:150:61
  48: <tonic::transport::service::router::RoutesFuture as core::future::future::Future>::poll
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tonic-0.7.1/src/transport/service/router.rs:90:36
  49: <tonic::transport::service::grpc_timeout::ResponseFuture<F> as core::future::future::Future>::poll
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tonic-0.7.1/src/transport/service/grpc_timeout.rs:88:38
  50: <tower::util::either::Either<A,B> as core::future::future::Future>::poll
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tower-0.4.12/src/util/either.rs:71:57
  51: <tonic::transport::server::recover_error::ResponseFuture<F> as core::future::future::Future>::poll
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tonic-0.7.1/src/transport/server/recover_error.rs:63:20
  52: <tonic::transport::server::SvcFuture<F> as core::future::future::Future>::poll
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tonic-0.7.1/src/transport/server/mod.rs:757:50
  53: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/core/src/future/future.rs:124:9
  54: hyper::proto::h2::server::H2Stream<F,B>::poll2
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.14.18/src/proto/h2/server.rs:448:37
  55: <hyper::proto::h2::server::H2Stream<F,B> as core::future::future::Future>::poll
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.14.18/src/proto/h2/server.rs:537:9
  56: tokio::runtime::task::core::CoreStage<T>::poll::{{closure}}
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.0/src/runtime/task/core.rs:165:17
  57: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.0/src/loom/std/unsafe_cell.rs:14:9
  58: tokio::runtime::task::core::CoreStage<T>::poll
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.0/src/runtime/task/core.rs:155:13
  59: tokio::runtime::task::harness::poll_future::{{closure}}
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.0/src/runtime/task/harness.rs:470:19
  60: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/core/src/panic/unwind_safe.rs:271:9
  61: std::panicking::try::do_call
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/std/src/panicking.rs:492:40
  62: __rust_try
  63: std::panicking::try
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/std/src/panicking.rs:456:19
  64: std::panic::catch_unwind
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/std/src/panic.rs:137:14
  65: tokio::runtime::task::harness::poll_future
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.0/src/runtime/task/harness.rs:458:18
  66: tokio::runtime::task::harness::Harness<T,S>::poll_inner
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.0/src/runtime/task/harness.rs:104:27
  67: tokio::runtime::task::harness::Harness<T,S>::poll
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.0/src/runtime/task/harness.rs:57:15
  68: tokio::runtime::task::raw::poll
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.0/src/runtime/task/raw.rs:144:5
  69: tokio::runtime::task::raw::RawTask::poll
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.0/src/runtime/task/raw.rs:84:18
  70: tokio::runtime::task::LocalNotified<S>::run
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.0/src/runtime/task/mod.rs:376:9
  71: tokio::runtime::thread_pool::worker::Context::run_task::{{closure}}
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.0/src/runtime/thread_pool/worker.rs:454:21
  72: tokio::coop::with_budget::{{closure}}
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.0/src/coop.rs:102:9
  73: std::thread::local::LocalKey<T>::try_with
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/std/src/thread/local.rs:442:16
  74: std::thread::local::LocalKey<T>::with
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/std/src/thread/local.rs:418:9
  75: tokio::coop::with_budget
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.0/src/coop.rs:95:5
  76: tokio::coop::budget
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.0/src/coop.rs:72:5
  77: tokio::runtime::thread_pool::worker::Context::run_task
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.0/src/runtime/thread_pool/worker.rs:430:9
  78: tokio::runtime::thread_pool::worker::Context::run
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.0/src/runtime/thread_pool/worker.rs:397:24
  79: tokio::runtime::thread_pool::worker::run::{{closure}}
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.0/src/runtime/thread_pool/worker.rs:382:17
  80: tokio::macros::scoped_tls::ScopedKey<T>::set
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.0/src/macros/scoped_tls.rs:61:9
  81: tokio::runtime::thread_pool::worker::run
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.0/src/runtime/thread_pool/worker.rs:379:5
  82: tokio::runtime::thread_pool::worker::Launch::launch::{{closure}}
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.0/src/runtime/thread_pool/worker.rs:358:45
  83: <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.0/src/runtime/blocking/task.rs:42:21
  84: tokio::runtime::task::core::CoreStage<T>::poll::{{closure}}
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.0/src/runtime/task/core.rs:165:17
  85: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.0/src/loom/std/unsafe_cell.rs:14:9
  86: tokio::runtime::task::core::CoreStage<T>::poll
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.0/src/runtime/task/core.rs:155:13
  87: tokio::runtime::task::harness::poll_future::{{closure}}
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.0/src/runtime/task/harness.rs:470:19
  88: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/core/src/panic/unwind_safe.rs:271:9
  89: std::panicking::try::do_call
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/std/src/panicking.rs:492:40
  90: __rust_try
  91: std::panicking::try
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/std/src/panicking.rs:456:19
  92: std::panic::catch_unwind
             at /rustc/f4a7ce997a1d7546d2b737f8b87d36907bcea2ad/library/std/src/panic.rs:137:14
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
Fri Apr 29 09:46:28 UTC 2022 [risedev]: Program exited with 134
```

🤣

**Comment by wangrunji0408** [comment]
Bug fixed in bd593d5. Thanks to `git bisect`.

**Pull request merged**